### PR TITLE
feat: rewrite and move to lightning

### PR DESCRIPTION
**Pull request by stephantul**
This PR removes all custom training code and rewrites the whole thing to Lightning. It makes use of all the primitives we defined in the model2vec codebase for classification and other tasks.

It also adds wandb tracking.

### CODE REVIEW

File: tokenlearn/losses.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        :param X: Tensor of shape (batch_size, d)
+        :param Y: Tensor of shape (batch_size, d)
+        :return: Scalar loss
+        """
+        cos_sim = cosine_similarity(X, Y)
+        loss = 1 - cos_sim.mean()
+        return loss
+
+
+class PairwiseRankingLoss(nn.Module):
```
Comment (Pringled) [inline tokenlearn/losses.py:47]: I think this is a generic contrastive loss, not pairwise ranking?

File: tokenlearn/train.py [hunk 2]
```diff
@@ -103,11 +63,24 @@ def main() -> None:
         default=256,
         help="Number of dimensions to reduce the target embeddings to using PCA.",
     )
+    parser.add_argument("--use-wandb", action="store_true", help="Use Weights & Biases for logging.")
+    parser.add_argument("--project-name", type=str, default="tokenlearn", help="Weights & Biases project name.")
+    parser.add_argument("--run-name", type=str, help="Weights & Biases run name.")
```
Comment (Pringled) [inline tokenlearn/train.py:68]: IMO this should not be a required arg, wandb will just create a default run name which is fine by default. I do like having it configurable though

File: tokenlearn/model.py [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+        else:
+            wandb_logger = None
+
+        # If the dataset is small, we check the validation set every epoch.
+        # If the dataset is large, we check the validation set every 250 batches.
+        if n_train_batches < 250:
+            val_check_interval = None
+            check_val_every_epoch = 1
+        else:
+            val_check_interval = max(250, 2 * len(val_dataset) // batch_size)
```
Comment (Pringled) [inline tokenlearn/model.py:137]: Hmm, what if you use a small bs (because of hardware constraints or for other reasons)? Then you would validate really often which can slow down training. Would it not make more sense to e.g. validate N times per epoch?

File: tokenlearn/model.py [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+        callbacks: list[Callback] = []
+        checkpoint_callback = ModelCheckpoint(monitor="val_loss", mode="min", save_top_k=1)
+        callbacks.append(checkpoint_callback)
+        if early_stopping_patience is not None:
+            callback = EarlyStopping(monitor="val_loss", mode="min", patience=early_stopping_patience)
+            callbacks.append(callback)
+        if use_wandb:
+            wandb_logger = WandbLogger(project=project_name, name=run_name)
+            lr_monitor = LearningRateMonitor(logging_interval="step")
+            callbacks.append(lr_monitor)
```
Comment (Pringled) [inline tokenlearn/model.py:127]: I think it would also be nice to log the model artifact dir (see https://docs.wandb.ai/guides/track/log/log-models/), much easier to keep things organised that way. Doesn't have to be in this PR though

File: tokenlearn/model.py [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+        val_dataset = self._prepare_dataset(validation_texts, validation_labels)
+
+        c = _FineTuningLightningModule(self, learning_rate=learning_rate, loss_name=self.loss)
+
+        n_train_batches = len(train_dataset) // batch_size
+        callbacks: list[Callback] = []
+        checkpoint_callback = ModelCheckpoint(monitor="val_loss", mode="min", save_top_k=1)
+        callbacks.append(checkpoint_callback)
+        if early_stopping_patience is not None:
+            callback = EarlyStopping(monitor="val_loss", mode="min", patience=early_stopping_patience)
```
Comment (Pringled) [inline tokenlearn/model.py:122]: IMO min_delta should also be configurable (but defaulted like early_stopping_patience); I usually tweak that a bit

File: tokenlearn/model.py [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+
+        state_dict = {}
+        for weight_name, weight in best_model_weights["state_dict"].items():
+            state_dict[weight_name.removeprefix("model.")] = weight
+
+        self.load_state_dict(state_dict)
+        self.eval()
+        return self
+
+    def _prepare_dataset(self, X: list[str], y: torch.Tensor, max_length: int = 512) -> TextDataset:
```
Comment (Pringled) [inline tokenlearn/model.py:169]: Random thought that has nothing to do with this PR but do you think our performance would increase for longer sequences if we use a larger max_length with a base model that supports a context length > 512? I.e. we try to mimic the mean embedding produced by the base model (and optimize our token embeddings for that task), so I would assume we'd get better at that as well? 

And another thought: we currently don't cap the texts in our featurize before encoding; so if we use a model that has sequence length 2048, we create a mean + text of that length, and then we cap it here to a lower max_length even though we have a mean vector based on a longer sequence length. Should that not be aligned somehow?

File: tokenlearn/train.py [hunk 7]
```diff
@@ -103,11 +63,24 @@ def main() -> None:
...
+    parser.add_argument(
+        "--limit-samples", type=int, default=None, help="Limit the number of samples to use for training."
+    )
+    parser.add_argument(
+        "--loss",
+        default="ranking",
+        choices=Loss.__members__.values(),
+        help="The loss function to use for training.",
+    )
+    parser.add_argument("--batch-size", type=int, default=_DEFAULT_BATCH_SIZE, help="Batch size for training.")
```
Comment (Pringled) [inline tokenlearn/train.py:78]: Would also make LR configurable here

**Review by Pringled — CHANGES_REQUESTED** [review]
Nice, much cleaner. Some thoughts and comments

File: tokenlearn/train.py (hunk 7 above)
Comment (stephantul) [reply tokenlearn/train.py:80]: Done

File: tokenlearn/model.py (hunk 6 above)
Comment (stephantul) [reply tokenlearn/model.py:169]: You are right, this is weird right now for models with longer sequence lengths. The 512 is aligned with model2vec itself, which has a max_length of 512, and which is not stored. So we could make that configurable there, and then later on make it configurable here.

File: tokenlearn/model.py (hunk 3 above)
Comment (stephantul) [reply tokenlearn/model.py:137]: This code is just copied from the model2vec repo, it works well but isn't really needed here, I agree. I don't like validate N times per epoch because that can really slow down training for smaller sets, which is why we use this in model2vec, I guess.

File: tokenlearn/train.py (hunk 2 above)
Comment (stephantul) [reply tokenlearn/train.py:69]: This is not a required argument

Comment (Pringled) [reply tokenlearn/train.py:69]: Oh nvm I misread 🤦

File: tokenlearn/model.py (hunk 3 above)
Comment (Pringled) [reply tokenlearn/model.py:137]: Yeah makes sense. It's not a big deal so fine to leave it as is
